lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -795,12 +795,23 @@ getConformanceAndConcreteType(FullApplySite AI,
     if (Requirement == Protocol) {
       return std::make_tuple(Conformance, ConcreteType, ConcreteTypeDef);
     }
-    if (Requirement->inheritsFrom(Protocol)) {
-      // If Requirement != Protocol, then the abstract conformance cannot be used
-      // as is and we need to create a proper conformance.
-      return std::make_tuple(Conformance.getInherited(Protocol), ConcreteType,
-                             ConcreteTypeDef);
-    }
+    // If Requirement != Protocol, then the abstract conformance cannot be
+    // used as is and we need to create a proper conformance.
+    // FIXME: We can handle only direct inheritance at the moment due to some
+    // limitations of the init_existential_* instructions representation.
+    // Once these instructions start using generic signatures instead of
```

Comment:
I think it's OK for the instruction to store just a list of conformances, but we should add some utilities to make this substitution map easy to construct. I'm thinking of the following:

- Add a WitnessMethodInst::getExistentialType() method, which returns a protocol composition type consisting of all the protocol requirements in the conformance array.
- Add ASTContext::getExistentialSignature(), which takes an existential type, and builds and caches a generic signature with a single generic parameter, with the requirements derived from TypeBase::getExistentialLayout().
- Then you need a utility method somewhere, perhaps in ExistentialLayout, which takes the concrete type, the array of conformances, and returns a SubstitutionMap from this.
- Now you can call lookupConformance() on the SubstitutionMap.

I think getExistentialSignature() will come in handy if we ever do generalized existentials. Also, we could model open existential types as having a generic environment with the existential signature. This would eliminate opened existentials as a special case in a few places (but we'd still have to model the data dependency aspect of them).

Reply:
@slavapestov The idea to store the signature in the instruction belongs to @DougGregor. He thought it would be better to do it instead of storing just the list of conformances.

Reply:
Ok. You'll still need to store the conformances separately (or use a SubstitutionList) since the generic signature only states the *requirements*.

Either way, we'll want a utility in ASTContext to build a signature from an existential type -- the only open question is do we call this from the SIL combiner here, or store it in the instruction itself during SILGen.